### PR TITLE
DALF Reader: Refactor decoder environment

### PR DESCRIPTION
**Pull request by hurryabit**
Currently, the module to be decoded is part of the decoder environment.
This turned out to be unpleasant during my attempts to implement
interning for types since decoding the type interning table does not
happen in the context of a module.

This PR moves the module out of the decoder environment and passes it
to the decoder function for modules directly. Unfornatunately, we still
need to keep the module name in the environment since that is used as a
default when decoding location information. We solve this problem by
making the module name in the environment optional and always filling
it in when decoding in actual module.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

---
This change is [Reviewable](https://reviewable.io/reviews/digital-asset/daml/7891)

### CODE REVIEW

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 1]
```diff
@@ -416,20 +418,20 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
...
           if (loc.hasModule)
-            decodeModuleRef(loc.getModule)
+            Some(decodeModuleRef(loc.getModule))
           else
-            (packageId, moduleName)
-
-        val range = loc.getRange
-        Some(
+            optModuleName.map((packageId, _))
+        optModuleRef.map(moduleRef => {
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:426]:
```suggestion
        optModuleRef.map{ case (pkgId, module) => 
```

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 2]
```diff
@@ -416,20 +418,20 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
...
+          val range = loc.getRange
           Location(
-            pkgId,
-            module,
+            moduleRef._1,
+            moduleRef._2,
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:430]:
```suggestion
            pkgId,
            module,
```

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 3]
```diff
@@ -416,20 +418,20 @@ private[archive] class DecodeV1(minor: LV.Minor) extends Decode.OfPackage[PLF.Pa
...
             definition,
             (range.getStartLine, range.getStartCol),
-            (range.getEndLine, range.getEndCol)))
+            (range.getEndLine, range.getEndCol))
+        })
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:434]:
```suggestion
        }
```

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good to me.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala (hunk 1 above)
Comment (hurryabit) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:426]: Ah, that's how it's done. I was trying to do that, but couldn't convince the type checker. Apparently, you can't the argument to `map` in parentheses `(...)` if it's already wrapped in curly braces `{...}`. Thanks.

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 4]
```diff
@@ -45,28 +45,17 @@ class DecodeV1Spec
   val dummyModuleDName = DamlLf1.DottedName.newBuilder().addSegments(dummyModuleStr).build()
   val dummyModuleName = DottedName.assertFromString(dummyModuleStr)
 
-  private def dummyModule(version: LV.Minor, interningIdx: Int) = {
```
Comment (hurryabit) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:48]: @remyhaemmerle-da I forgot to fix the tests on the first round. They got actually simpler. Yay!

Comment (remyhaemmerle-da) [reply daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:48]: :+1:

**Pull request merged**